### ISSUE TITLE
Spack broken on Slurm v5 due to hardcoded 300s custom script timeout

### ISSUE THREAD

**Issue by m4r1k** [GoogleCloudPlatform/hpc-toolkit#536]
### Describe the bug

When deploying a blueprint that includes long running builds, SlurmGCP v5 has an hardcoded 300 seconds timeout and it fails.
https://github.com/SchedMD/slurm-gcp/blob/v5.1.0/scripts/setup.py#L564

Find please at the end of this report the blueprint

Follows the resulting error log (taken from /slurm/scripts/setup.log)
```text
2022-09-09 21:35:17,466 DEBUG: run_custom_scripts: custom scripts to run: /slurm/custom_scripts/(login_6onjpjm6.d/ghpc_startup.sh)
2022-09-09 21:35:17,466 INFO: running script ghpc_startup.sh
2022-09-09 21:40:17,488 ERROR: TimeoutExpired:
    command=/slurm/custom_scripts/login_6onjpjm6.d/ghpc_startup.sh
    timeout=300
    stdout:
```

### Steps to reproduce

1. `ghpc create <yaml blueprint file>`
2. `terraform -chdir=internal-hpc/primary init`
3. `terraform -chdir=internal-hpc/primary validate`
4. `terraform -chdir=internal-hpc/primary apply`

### Expected behavior

Just as much in Slurm v4 I was able to build stuff that takes long time (like Gromacs and GCC), I should be able to achieve the same in Slurm v5.

### Actual behavior

I cannot have a fully working setup that includes packages that takes long to compile.

### Version (`ghpc --version`)
`ghpc version v1.4.1`

### Blueprint
```yaml
---
blueprint_name: hpc-cluster
vars:
  project_id: internal-hpc
  deployment_name: internal-hpc
  region: europe-west4
  zone: europe-west4-b

# Documentation for each of the modules used below can be found at
# https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/modules/README.md

deployment_groups:
- group: primary
  modules:
  - source: modules/network/vpc
    kind: terraform
    settings:
      mtu: 8896
    id: network1

  ## Filesystems
  - source: modules/file-system/filestore
    kind: terraform
    id: appsfs
    use: [network1]
    settings:
      local_mount: /sw
      filestore_tier: BASIC_HDD
      size_gb: 1024

  - source: modules/file-system/filestore
    kind: terraform
    id: homefs
    use: [network1]
    settings:
      local_mount: /home
      filestore_tier: BASIC_SSD
      size_gb: 3072

  ## Spack Install Scripts
  - source: community/modules/scripts/spack-install
    kind: terraform
    id: spack
    settings:
      install_dir: /sw/spack
      spack_url: https://github.com/spack/spack
      spack_ref: v0.18.1
      log_file: /var/log/spack.log
      configs:
      - type: single-config
        scope: defaults
        content: "config:build_stage:/sw/spack/spack-stage"
      - type: file
        scope: defaults
        content: |
          modules:
            default:
              tcl:
                hash_length: 0
                all:
                  conflict:
                    - '{name}'
                projections:
                  all: '{name}/{version}-{compiler.name}-{compiler.version}'
      compilers:
      - gcc@12.1.0 target=x86_64
      packages:
      - intel-mpi@2018.4.274%gcc@12.1.0 target=x86_64_v3
      - intel-oneapi-mpi@2021.6.0%gcc@12.1.0 target=x86_64_v3
      ## https://github.com/spack/spack/blob/v0.18.1/lib/spack/external/archspec/json/cpu/microarchitectures.json

  - source: modules/scripts/startup-script
    kind: terraform
    id: spack-startup
    settings:
      runners:
      - type: shell
        source: modules/startup-script/examples/install_ansible.sh
        destination: install_ansible.sh
      - $(spack.install_spack_deps_runner)
      - $(spack.install_spack_runner)

  - source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    kind: terraform
    id: c2d
    use:
    - network1
    - homefs
    - appsfs
    settings:
      partition_name: c2d
      node_count_dynamic_max: 10
      enable_placement: true
      machine_type: "c2d-highcpu-112"
      disk_type: "pd-standard"
      disable_smt: true
      bandwidth_tier: "gvnic_enabled"
      is_default: true

  - source: community/modules/compute/schedmd-slurm-gcp-v5-partition
    kind: terraform
    id: c2
    use:
    - network1
    - homefs
    - appsfs
    settings:
      partition_name: c2
      node_count_dynamic_max: 20
      enable_placement: true
      machine_type: "c2-standard-60"
      disk_type: "pd-standard"
      disable_smt: true
      bandwidth_tier: "gvnic_enabled"

  - source: community/modules/scheduler/schedmd-slurm-gcp-v5-controller
    kind: terraform
    id: slurm_controller
    use:
    - network1
    - homefs
    - appsfs
    - c2d
    - c2
    settings:
      disk_type: "pd-ssd"
      disk_size_gb: 50
      machine_type: "t2d-standard-2"
      disable_controller_public_ips: true

  - source: community/modules/scheduler/schedmd-slurm-gcp-v5-login
    kind: terraform
    id: slurm_login
    use:
    - network1
    - homefs
    - appsfs
    - slurm_controller
    - spack-startup
    settings:
      disk_type: "pd-ssd"
      disk_size_gb: 50
      machine_type: "t2d-standard-32"

  - source: modules/monitoring/dashboard
    kind: terraform
    id: hpc_dashboard
```

**Comment by cboneti** [comment]
Hi Federico, 

Thank you for reporting this. We are aware of the problem and we are working to find the best solution. 
Right now, we are considering changing most of our examples so that they don't build on the login node but, instead, use a "spack builder" node. It has the advantage that it would allow slurm boot independently and that it would be easy to add or remove packages to the yaml. 

Please let me know if this is an urgent issue for you and we can prioritize having a fully tested working example. 

Regards, 

Carlos

**Comment by m4r1k** [comment]
Hey Carlos,

I think leveraging a spack builder node would go a long way for several reasons:
1) having the builds running on the login nodes results in either long building times or an expensive idling object once done
2) decoupling compilers flags from the machine types used for the login nodes (e.g. building and tuning on C2 while running a relatively cheap yet different micro-arch on t2d on the login node)
3) optimize for different architectures (let's say c2 and c2d) through multiple building nodes

so, totally +1 from me. And yes, for the time being, removing the examples makes sense :-)

**Comment by tpdownes** [comment]
@m4r1k please follow https://github.com/GoogleCloudPlatform/hpc-toolkit/pull/558 for progress

**Comment by tpdownes** [comment]
#558 has been merged to the `develop` branch and provides a model for installing Spack on a build VM. Right now, the example installs Spack but no actual Spack packages. We plan on improving that as time goes on, but this is a good first example to follow.

**Comment by tpdownes** [comment]
We have also added a [specific troubleshooting section](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/develop/docs/slurm-troubleshooting.md#timeout-error--startup-script-failure-slurm-v5) on this issue.

**Comment by tpdownes** [comment]
The blueprint using a build VM was released in v1.6.0 of the Toolkit. The upcoming v1.7.0 release improves upon the blueprint by using Spack to install an application optimized for the AMD platform.

https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/examples/AMD/hpc-cluster-amd-slurmv5.yaml
